nexosim/tests/integration/simulation_halt.rs

```diff
@@ -0,0 +1,91 @@
+//! Simulation pause and resume
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::time::{Duration, Instant};
+
+use nexosim::model::{Context, InitializedModel, Model};
+use nexosim::ports::{EventQueue, Output};
+use nexosim::simulation::{ExecutionError, Mailbox, SimInit, SimulationError};
+use nexosim::time::{MonotonicTime, SystemClock};
+
+struct RecurringModel {
+    pub message: Output<()>,
+    delay: Duration,
+}
+impl RecurringModel {
+    fn new(delay: Duration) -> Self {
+        Self {
+            delay,
+            message: Output::new(),
+        }
+    }
+    async fn process(&mut self) {
+        self.message.send(()).await;
+    }
+}
+impl Model for RecurringModel {
+    async fn init(self, cx: &mut Context<Self>) -> InitializedModel<Self> {
+        cx.schedule_periodic_event(self.delay, self.delay, RecurringModel::process, ())
+            .unwrap();
+        self.into()
+    }
+}
+
+#[test]
+fn halt_and_resume() -> Result<(), SimulationError> {
+    let mut model = RecurringModel::new(Duration::from_secs(2));
+    let mailbox = Mailbox::new();
+
+    let message = EventQueue::new();
+    model.message.connect_sink(&message);
+    let mut message = message.into_reader();
+
+    let t0 = MonotonicTime::EPOCH;
+
+    let (simu, mut scheduler) = SimInit::new()
+        .add_model(model, mailbox, "timed_model")
+        .set_clock(SystemClock::from_instant(t0, Instant::now()))
+        .init(t0)?;
+
+    let simulation = Arc::new(Mutex::new(simu));
+    let spawned_simulation = simulation.clone();
+
+    thread::spawn(move || spawned_simulation.lock().unwrap().step_unbounded());
+
+    thread::sleep(Duration::from_secs(1));
+
+    scheduler.halt();
+    thread::sleep(Duration::from_secs(1));
+
+    // assert that the step has completed, even though `pause` was called before it's scheduled time
+    assert!(message.next().is_some());
+
+    thread::sleep(Duration::from_secs(3));
+
+    // halted - no new messages
+    assert!(message.next().is_none());
+
+    {
+        let mut s = simulation.lock().unwrap();
+        let t = s.time();
+        s.reset_clock(SystemClock::from_instant(t, Instant::now()));
```

Comment:
You can use autoclock, I think

Reply:
I can, but the reset behaviour would be slighlty different. I'd resume on the next scheduled event (simulation-time) rather than on where I halted. I think from the simulation point of view it does not matter, but the `assert!(message.next().is_none());` would have to be removed (as it'd fail).

Reply:
Ok, I agree.